File: 2018/day08.py
import fileinput


class TreeNode:
    def __init__(self, numbers):
        num_children = next(numbers)
        num_metadata = next(numbers)
        self.children = [TreeNode(numbers) for i in range(num_children)]
        self.metadata = [next(numbers) for i in range(num_metadata)]

    def metadata_sum(self):
        return sum(self.metadata) + sum(
            child.metadata_sum() for child in self.children
        )

    def value(self):
        if self.children:
            return sum(self.children[index - 1].value()
                       for index in self.metadata
                       if 0 < index <= len(self.children))
        else:
           return sum(self.metadata)


def p1(data):
    return TreeNode(map(int, data.split())).metadata_sum()


def p2(data):
    return TreeNode(map(int, data.split())).value()


if __name__ == "__main__":
    data = fileinput.input()[0]
    print("Part 1: %s" % p1(data))
    print("Part 2: %s" % p2(data))

File: 2018/cpp/day08.cc
#include <algorithm>
#include <memory>
#include <numeric>
#include <string>
#include <vector>

#include "utils.h"

namespace {
using std::string;
using std::vector;
}  // namespace

class TreeNode {
 private:
  const vector<TreeNode> children;
  const vector<int> metadata;

  TreeNode(const vector<TreeNode> children, const vector<int> metadata)
      : children(children), metadata(metadata){};

  static TreeNode from_iterator_pointer(vector<int>::const_iterator *it) {
    vector<TreeNode> children;
    vector<int> metadata;
    int num_children = *(*it)++;
    int num_metadata = *(*it)++;
    children.reserve(num_children);
    metadata.reserve(num_metadata);
    for (int i = 0; i < num_children; ++i) {
      children.push_back(TreeNode::from_iterator_pointer(it));
    }
    for (int i = 0; i < num_metadata; ++i) {
      metadata.push_back(*(*it)++);
    }

    return TreeNode(children, metadata);
  }

 public:
  static TreeNode from_numbers(const vector<int> &numbers) {
    auto it = numbers.begin();
    return TreeNode::from_iterator_pointer(&it);
  }

  int metadata_sum() const {
    int total = std::accumulate(metadata.begin(), metadata.end(), 0);
    for (const auto &child : children) {
      total += child.metadata_sum();
    }

    return total;
  }

  int value() const {
    if (children.size()) {
      int total = 0;
      for (int i : metadata) {
        if (0 < i && i <= children.size()) {
          total += children[i - 1].value();
        }
      }
      return total;
    } else {
      return metadata_sum();
    }
  }
};

int p1(vector<int> numbers) {
  return TreeNode::from_numbers(numbers).metadata_sum();
}

int p2(vector<int> numbers) { return TreeNode::from_numbers(numbers).value(); }

int main() {
  vector<string> lines = aoc::utils::getlines();
  vector<string> strings = aoc::utils::split(lines[0], ' ');

  vector<int> numbers;
  std::transform(strings.begin(), strings.end(), std::back_inserter(numbers),
                 [](string &s) { return std::stoi(s); });

  std::cout << "Part 1: " << p1(numbers) << std::endl;
  std::cout << "Part 2: " << p2(numbers) << std::endl;
}
